### PR TITLE
fix(ecmascript): add argument validation for NewExpression side-effect detection

### PR DESCRIPTION
**Pull request by Dunqing**
## Summary

`is_pure_constructor` treated many global constructors as unconditionally side-effect-free regardless of arguments. Cross-tool research (ES spec, esbuild, Terser, Rollup, SWC) confirmed this was the most permissive of all bundlers/minifiers and introduced real bugs.

### Problem

These constructors perform observable operations on object arguments but were treated as pure:

| Constructor | Side effect | ES Spec operation |
|---|---|---|
| `new String(obj)` | Calls `toString()`/`valueOf()` | `ToString` -> `ToPrimitive` |
| `new Number(obj)` | Calls `valueOf()`/`toString()`, throws on Symbol | `ToNumeric` -> `ToPrimitive` |
| `new Date(obj)` | Calls `[Symbol.toPrimitive]`/`valueOf()`/`toString()` | `ToPrimitive` |
| `new TypedArray(obj)` | Calls `[Symbol.iterator]` or `.length` getter | `GetMethod(@@iterator)` |
| `new ArrayBuffer(obj)` | Calls `valueOf()`/`toString()` | `ToIndex` -> `ToNumber` |

### Cross-tool comparison

| Constructor | esbuild | Terser | Rollup | SWC | Oxc (before) |
|---|---|---|---|---|---|
| `new String(obj)` | impure | impure (`unsafe` only) | pure | pure | **pure** |
| `new Number(obj)` | impure | impure (`unsafe` only) | pure | pure | **pure** |
| `new Date(obj)` | impure | impure (`unsafe` only) | pure | impure | **pure** |
| `new TypedArray(obj)` | impure | impure | pure | impure | **pure** |

### Solution

- Split `is_pure_constructor` into `is_unconditionally_pure_constructor` (Object, Boolean, Error types) and `is_typed_array_constructor`
- Add per-constructor argument validation in `NewExpression::may_have_side_effects` using the same `ToPrimitive`-based pattern already used by `CallExpression`
- Three shared helpers: `new_expr_with_to_string_check`, `new_expr_with_to_number_check`, `new_expr_with_to_primitive_check`

### Also fixes

- Typo `"BitUint64Array"` -> `"BigUint64Array"` in `is_known_global_constructor`
- `is_pure_collection_constructor` now verifies `undefined` is actually global (not shadowed)

## Test plan

- Added 52 new test cases covering primitive args (safe), unknown/object args (side-effectful), and edge cases (`{toString(){}}`, `{valueOf(){}}`)
- All existing tests continue to pass
- Verified no regressions in Rolldown (1639/1639 tests pass with this change patched in)

### CODE REVIEW

**Comment by Dunqing** [comment]
* **#20420** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/20420?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#20401** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/20401?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#20395** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/20395?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/oxc-project/oxc/20395?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#20383** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/20383?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent changes, fast-track this PR to the front of the merge queue

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.com/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=20395).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: crates/oxc_ecmascript/src/side_effects/expressions.rs [hunk 1]
```diff
@@ -293,14 +293,23 @@ fn is_pure_call(name: &str) -> bool {
             | "SyntaxError" | "TypeError" | "URIError" | "Number" | "Object" | "String" | "Symbol")
 }
 
+/// Constructors that are unconditionally side-effect-free with any arguments.
+///
+/// - `Object`: wraps/returns any argument, no coercion
+/// - `Boolean`: `ToBoolean` is a purely internal operation, no user code
+/// - Error types: just create error objects, no observable coercion
```
Comment (sapphi-red) [inline crates/oxc_ecmascript/src/side_effects/expressions.rs:300]: `new Error(msg)` will call ToString for `msg`.

File: crates/oxc_ecmascript/src/side_effects/expressions.rs [hunk 2]
```diff
@@ -1025,6 +1042,51 @@ impl<'a> MayHaveSideEffects<'a> for CallExpression<'a> {
     }
 }
 
+/// Pure if 0 args or first arg's `ToPrimitive` yields a known string.
+/// Used for `new String(arg)` where `ToString` calls `ToPrimitive(input, string)`.
```
Comment (sapphi-red) [inline crates/oxc_ecmascript/src/side_effects/expressions.rs:1052]: I think we don't need to treat these as sideeffects. We have this assumption: https://github.com/oxc-project/oxc/blob/main/crates/oxc_minifier/docs/ASSUMPTIONS.md#coercion-methods-are-pure
(other places may throw because Symbol might be passed to `ToString`, but in this case that never happens)

File: crates/oxc_ecmascript/src/side_effects/expressions.rs [hunk 3]
```diff
@@ -1025,6 +1042,51 @@ impl<'a> MayHaveSideEffects<'a> for CallExpression<'a> {
...
+fn new_expr_with_to_primitive_check<'a>(
+    expr: &NewExpression<'a>,
+    ctx: &impl MayHaveSideEffectsContext<'a>,
+) -> bool {
+    if expr.arguments.iter().any(|e| e.may_have_side_effects(ctx)) {
+        return true;
+    }
+    expr.arguments.first().is_some_and(|arg| {
+        arg.as_expression()
+            .is_none_or(|e| matches!(e.to_primitive(ctx), ToPrimitiveResult::Undetermined))
```
Comment (sapphi-red) [inline crates/oxc_ecmascript/src/side_effects/expressions.rs:1092]: If it's converted to BigInt or Symbol, `ToNumber` will throw an error.
If it's an Object, `TypedArray` may throw an error because `@@iterator` may throw an error.

**Review by sapphi-red — CHANGES_REQUESTED** [review]

File: crates/oxc_ecmascript/src/side_effects/expressions.rs (hunk 1 above)
Comment (Dunqing) [reply crates/oxc_ecmascript/src/side_effects/expressions.rs:300]: Addressed in #20420 — updated the doc comment to note that `new Error(msg)` calls `ToString(msg)`, covered by the coercion assumption.

File: crates/oxc_ecmascript/src/side_effects/expressions.rs (hunk 2 above)
Comment (Dunqing) [reply crates/oxc_ecmascript/src/side_effects/expressions.rs:1052]: Addressed in #20420 — removed the `new_expr_with_to_string_check` helper and made `String`, `Number`, `Date`, `ArrayBuffer` unconditionally pure constructors per the coercion assumption. Also simplified `String()`, `Number()`, `Symbol()` call expressions the same way.

File: crates/oxc_ecmascript/src/side_effects/expressions.rs (hunk 3 above)
Comment (Dunqing) [reply crates/oxc_ecmascript/src/side_effects/expressions.rs:1092]: Addressed in #20420 — TypedArray now uses `value_type()` instead of `to_primitive()`. Only known safe primitives (Number, String, Boolean, Null, Undefined) are accepted; Object args (@@iterator) and BigInt args (ToNumber throws) are correctly treated as side-effectful.

**Review by sapphi-red — APPROVED** [review]
Fixed in #20420
